docs/code_examples/aec_rps.py

```diff
@@ -60,29 +60,36 @@ def __init__(self, render_mode=None):
         The init method takes in environment arguments and
          should define the following attributes:
         - possible_agents
-        - action_spaces
-        - observation_spaces
+        - render_mode
+
+        Note: the action_spaces and observation_spaces attributes are deprecated.
+        Spaces must be defined in the action_space() and observation_space() methods.
+
         These attributes should not be changed after initialization.
         """
         self.possible_agents = ["player_" + str(r) for r in range(2)]
+
+        # optional: a mapping between agent name and ID
         self.agent_name_mapping = dict(
             zip(self.possible_agents, list(range(len(self.possible_agents))))
         )
 
-        # gymnasium spaces are defined and documented here: https://gymnasium.farama.org/api/spaces/
+        # optional: explicitly define spaces as dicts (deprecated)
         self._action_spaces = {agent: Discrete(3) for agent in self.possible_agents}
         self._observation_spaces = {
             agent: Discrete(4) for agent in self.possible_agents
         }
         self.render_mode = render_mode
 
+    # observation space must be defined here
     # this cache ensures that same space object is returned for the same agent
-    # allows action space seeding to work as expected
+    # allows observation and action space seeding to work as expected
```

Comment:
"This allows observation and action spaces to be memoized, reducing clock cycles required to get each agent's action and observation space. If your spaces change over time, do not use this line."